File: app/src/main/java/vn/unlimit/vpngate/models/VPNGateConnection.java
package vn.unlimit.vpngate.models;

import android.content.Context;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.Base64;

import java.text.DecimalFormat;

import vn.unlimit.vpngate.App;
import vn.unlimit.vpngate.R;
import vn.unlimit.vpngate.utils.DataUtil;

/**
 * Created by dongh on 14/01/2018.
 */

public class VPNGateConnection implements Parcelable {
    public static final Parcelable.Creator<VPNGateConnection> CREATOR
            = new Parcelable.Creator<VPNGateConnection>() {
        public VPNGateConnection createFromParcel(Parcel in) {
            return new VPNGateConnection(in);
        }

        public VPNGateConnection[] newArray(int size) {
            return new VPNGateConnection[size];
        }
    };
    //HostName,IP,Score,Ping,Speed,CountryLong,CountryShort,NumVpnSessions,Uptime,TotalUsers,TotalTraffic,logType,Operator,Message,OpenVPN_ConfigData_Base64
    private String hostName;
    private String ip;
    private int score;
    private int ping;
    private int speed;
    private String countryLong;
    private String countryShort;
    private int numVpnSession;
    private int uptime;
    private int totalUser;
    private long totalTraffic;
    private String logType;
    private String operator;
    private String message;
    private String openVpnConfigData;
    private int tcpPort;
    private int udpPort;
    private int isL2TPSupport;
    private int isSSTPSupport;

    private VPNGateConnection(Parcel in) {
        hostName = in.readString();
        ip = in.readString();
        score = in.readInt();
        ping = in.readInt();
        speed = in.readInt();
        countryLong = in.readString();
        countryShort = in.readString();
        numVpnSession = in.readInt();
        uptime = in.readInt();
        totalUser = in.readInt();
        totalTraffic = in.readLong();
        logType = in.readString();
        operator = in.readString();
        message = in.readString();
        openVpnConfigData = in.readString();
        tcpPort = in.readInt();
        udpPort = in.readInt();
        isL2TPSupport = in.readInt();
        isSSTPSupport = in.readInt();
    }

    //Empty constructor
    private VPNGateConnection() {

    }

    public static VPNGateConnection fromCsv(String csvLine) {
        String[] properties = csvLine.split(",");
        try {
            int index = 0;
            VPNGateConnection vpnGateConnection = new VPNGateConnection();
            vpnGateConnection.hostName = properties[index++];
            vpnGateConnection.ip = properties[index++];
            vpnGateConnection.score = Integer.parseInt(properties[index++]);
            vpnGateConnection.ping = Integer.parseInt(properties[index++]);
            vpnGateConnection.speed = Integer.parseInt(properties[index++]);
            vpnGateConnection.countryLong = properties[index++];
            vpnGateConnection.countryShort = properties[index++];
            vpnGateConnection.numVpnSession = Integer.parseInt(properties[index++]);
            vpnGateConnection.uptime = Integer.parseInt(properties[index++]);
            vpnGateConnection.totalUser = Integer.parseInt(properties[index++]);
            vpnGateConnection.totalTraffic = Long.parseLong(properties[index++]);
            vpnGateConnection.logType = properties[index++];
            vpnGateConnection.setOperator(properties[index++]);
            vpnGateConnection.message = properties[index++];
            vpnGateConnection.setOpenVpnConfigData(properties[index]);
            if (App.getInstance().getDataUtil().getBooleanSetting(DataUtil.INCLUDE_UDP_SERVER, true) && properties.length >= index + 2) {
                vpnGateConnection.tcpPort = Integer.parseInt(properties[++index]);
                vpnGateConnection.udpPort = Integer.parseInt(properties[++index]);
                if (properties.length > index + 1) {
                    vpnGateConnection.isL2TPSupport = Integer.parseInt(properties[++index]);
                }
                if (properties.length > index + 1) {
                    vpnGateConnection.isSSTPSupport = Integer.parseInt(properties[++index]);
                }
            } else {
                vpnGateConnection.tcpPort = 0;
                vpnGateConnection.udpPort = 0;
                vpnGateConnection.isL2TPSupport = 0;
                vpnGateConnection.isSSTPSupport = 0;
            }
            return vpnGateConnection;
        } catch (Exception e) {
            return null;
        }
    }

    public void writeToParcel(Parcel out, int flags) {
        out.writeString(hostName);
        out.writeString(ip);
        out.writeInt(score);
        out.writeInt(ping);
        out.writeInt(speed);
        out.writeString(countryLong);
        out.writeString(countryShort);
        out.writeInt(numVpnSession);
        out.writeInt(uptime);
        out.writeInt(totalUser);
        out.writeLong(totalTraffic);
        out.writeString(logType);
        out.writeString(operator);
        out.writeString(message);
        out.writeString(openVpnConfigData);
        out.writeInt(tcpPort);
        out.writeInt(udpPort);
        out.writeInt(isL2TPSupport);
        out.writeInt(isSSTPSupport);
    }

    private String decodeBase64(String base64str) {
        try {
            byte[] plainBytes = Base64.decode(base64str, 1);
            return new String(plainBytes);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public String getHostName() {
        return hostName;
    }

    public void setHostName(String hostName) {
        this.hostName = hostName;
    }

    public String getCalculateHostName() {
        return hostName + ".opengw.net";
    }

    public String getIp() {
        return ip;
    }

    public void setIp(String ip) {
        this.ip = ip;
    }

    public int getScore() {
        return score;
    }

    public void setScore(int score) {
        this.score = score;
    }

    public String getScoreAsString() {
        return String.valueOf(score);
    }

    public int getPing() {
        return ping;
    }

    public void setPing(int ping) {
        this.ping = ping;
    }

    public String getPingAsString() {
        return ping + "";
    }

    public String getCountryLong() {
        return countryLong;
    }

    public void setCountryLong(String countryLong) {
        this.countryLong = countryLong;
    }

    public String getCountryShort() {
        return countryShort;
    }

    public void setCountryShort(String countryShort) {
        this.countryShort = countryShort;
    }

    public int getNumVpnSession() {
        return numVpnSession;
    }

    public void setNumVpnSession(int numVpnSession) {
        this.numVpnSession = numVpnSession;
    }

    public String getNumVpnSessionAsString() {
        return numVpnSession + "";
    }

    public int getTotalUser() {
        return totalUser;
    }

    public void setTotalUser(int totalUser) {
        this.totalUser = totalUser;
    }

    public String getLogType() {
        return logType;
    }

    public void setLogType(String logType) {
        this.logType = logType;
    }

    public String getOperator() {
        return operator;
    }

    public void setOperator(String operator) {
        operator = operator.replace("'s owner", "");
        this.operator = operator;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public String getOpenVpnConfigData() {
        String openVpnConfigDataTmp = openVpnConfigData;
        if (App.getInstance().getDataUtil().getBooleanSetting(DataUtil.USE_DOMAIN_TO_CONNECT, false)) {
            String dDomain = hostName + ".opengw.net";
            openVpnConfigDataTmp = openVpnConfigDataTmp.replace(ip, dDomain);

        }
        return openVpnConfigDataTmp;
    }

    public String getOpenVpnConfigDataUdp() {
        String openVpnConfigDataUdp = openVpnConfigData;
        if (this.tcpPort > 0) {
            // Current config is config for tcp need for udp
            openVpnConfigDataUdp = openVpnConfigDataUdp
                    .replace("proto tcp", "proto udp")
                    .replace("remote " + ip + " " + tcpPort, "remote " + ip + " " + udpPort);
        }
        if (App.getInstance().getDataUtil().getBooleanSetting(DataUtil.USE_DOMAIN_TO_CONNECT, false)) {
            String dDomain = hostName + ".opengw.net";
            openVpnConfigDataUdp = openVpnConfigDataUdp.replace(ip, dDomain);

        }
        // Current config is udp only
        return openVpnConfigDataUdp;
    }

    public void setOpenVpnConfigData(String openVpnConfigData) {
        this.openVpnConfigData = decodeBase64(openVpnConfigData);
    }

    public int getSpeed() {
        return speed;
    }

    public void setSpeed(int speed) {
        this.speed = speed;
    }

    public String getCalculateSpeed() {
        return round((double) speed / (1000 * 1000));
    }

    public int getUptime() {
        return uptime;
    }

    public void setUptime(int uptime) {
        this.uptime = uptime;
    }

    public long getTotalTraffic() {
        return totalTraffic;
    }

    public void setTotalTraffic(long totalTraffic) {
        this.totalTraffic = totalTraffic;
    }

    public String getCalculateTotalTraffic() {
        double inMB = (double) totalTraffic / (1000 * 1000);
        if (inMB < 1000) {
            return round(inMB) + " MB";
        }
        double inGB = inMB / 1000;
        if (inGB < 1000) {
            return round(inMB / 1000) + " GB";
        }
        return round(inGB / 1000) + " TB";
    }

    public String getCalculateUpTime(Context context) {
        //Display as second
        if (uptime < 60000) {
            return round(uptime / 1000) + " " + context.getResources().getString(R.string.seconds);
        }
        //Display as minute
        if (uptime < 3600000) {
            return Math.round((double) uptime / 60000) + " " + context.getResources().getString(R.string.minutes);
        }
        //Display as hours
        if (uptime < 3600000 * 24) {
            return round((double) uptime / 3600000) + " " + context.getResources().getString(R.string.hours);
        }
        return round((double) uptime / (24 * 3600000)) + " " + context.getResources().getString(R.string.days);
    }

    private String round(double value) {
        DecimalFormat df = new DecimalFormat("####0.###");
        return df.format(value);
    }

    public int describeContents() {
        return 0;
    }

    public String getName() {
        return this.getName(false);
    }

    public String getName(boolean useUdp) {
        String address = ip;
        if (App.getInstance().getDataUtil().getBooleanSetting(DataUtil.USE_DOMAIN_TO_CONNECT, false)) {
            address = hostName + ".opengw.net";
        }
        if (App.getInstance().getDataUtil().getBooleanSetting(DataUtil.INCLUDE_UDP_SERVER, true)) {
            if (tcpPort == 0 && udpPort == 0) {
                // Current profile from non udp but open status page with include udp option
                return String.format("%s[%s]", countryLong, address);
            }
            return String.format("%s[%s][%s]", countryLong, address, useUdp || tcpPort == 0 ? "UDP:" + udpPort : "TCP:" + tcpPort);
        }
        return String.format("%s[%s]", countryLong, address);
    }

    public int getTcpPort() {
        return tcpPort;
    }

    public void setTcpPort(int tcpPort) {
        this.tcpPort = tcpPort;
    }

    public int getUdpPort() {
        return udpPort;
    }

    public void setUdpPort(int udpPort) {
        this.udpPort = udpPort;
    }

    public boolean isL2TPSupport() {
        return isL2TPSupport == 1;
    }

    public  boolean isSSTPSupport() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && isSSTPSupport == 1;
    }
}

File: app/src/main/java/vn/unlimit/vpngate/models/VPNGateConnection.kt
package vn.unlimit.vpngate.models

import android.content.Context
import android.os.Build
import android.os.Parcel
import android.os.Parcelable
import android.util.Base64
import vn.unlimit.vpngate.App.Companion.instance
import vn.unlimit.vpngate.R
import vn.unlimit.vpngate.utils.DataUtil
import java.text.DecimalFormat

/**
 * Created by dongh on 14/01/2018.
 */
class VPNGateConnection : Parcelable {
    //HostName,IP,Score,Ping,Speed,CountryLong,CountryShort,NumVpnSessions,Uptime,TotalUsers,TotalTraffic,logType,Operator,Message,OpenVPN_ConfigData_Base64
    var hostName: String? = null
    var ip: String? = null
    var score = 0
    var ping = 0
    var speed = 0
    var countryLong: String? = null
    var countryShort: String? = null
    var numVpnSession = 0
    var uptime = 0
    var totalUser = 0
    var totalTraffic: Long = 0
    var logType: String? = null
    var operator: String? = null
    var message: String? = null
    var openVpnConfigData: String? = null
    var tcpPort = 0
    var udpPort = 0
    private var isL2TPSupport = 0
    private var isSSTPSupport = 0

    private constructor(`in`: Parcel) {
        hostName = `in`.readString()
        ip = `in`.readString()
        score = `in`.readInt()
        ping = `in`.readInt()
        speed = `in`.readInt()
        countryLong = `in`.readString()
        countryShort = `in`.readString()
        numVpnSession = `in`.readInt()
        uptime = `in`.readInt()
        totalUser = `in`.readInt()
        totalTraffic = `in`.readLong()
        logType = `in`.readString()
        operator = `in`.readString()
        message = `in`.readString()
        openVpnConfigData = `in`.readString()
        tcpPort = `in`.readInt()
        udpPort = `in`.readInt()
        isL2TPSupport = `in`.readInt()
        isSSTPSupport = `in`.readInt()
    }

    //Empty constructor
    constructor()

    override fun writeToParcel(out: Parcel, flags: Int) {
        out.writeString(hostName)
        out.writeString(ip)
        out.writeInt(score)
        out.writeInt(ping)
        out.writeInt(speed)
        out.writeString(countryLong)
        out.writeString(countryShort)
        out.writeInt(numVpnSession)
        out.writeInt(uptime)
        out.writeInt(totalUser)
        out.writeLong(totalTraffic)
        out.writeString(logType)
        out.writeString(operator)
        out.writeString(message)
        out.writeString(openVpnConfigData)
        out.writeInt(tcpPort)
        out.writeInt(udpPort)
        out.writeInt(isL2TPSupport)
        out.writeInt(isSSTPSupport)
    }

    private fun decodeBase64(base64str: String): String? {
        try {
            val plainBytes = Base64.decode(base64str, 1)
            return String(plainBytes)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    val calculateHostName: String
        get() = "$hostName.opengw.net"

    val scoreAsString: String
        get() = score.toString()

    val pingAsString: String
        get() = ping.toString() + ""

    val numVpnSessionAsString: String
        get() = numVpnSession.toString() + ""

    fun setOperatorString(operator: String) {
        var loperator = operator
        loperator = loperator.replace("'s owner", "")
        this.operator = loperator
    }

    fun getOpenVpnConfigDataString(): String? {
        var openVpnConfigDataTmp = openVpnConfigData
        if (instance!!.dataUtil!!.getBooleanSetting(DataUtil.USE_DOMAIN_TO_CONNECT, false)) {
            val dDomain = "$hostName.opengw.net"
            openVpnConfigDataTmp = openVpnConfigDataTmp!!.replace(ip!!, dDomain)
        }
        return openVpnConfigDataTmp
    }

    fun setOpenVpnConfigDataString(openVpnConfigData: String) {
        this.openVpnConfigData = decodeBase64(openVpnConfigData)
    }

    val openVpnConfigDataUdp: String?
        get() {
            var openVpnConfigDataUdp = openVpnConfigData
            if (this.tcpPort > 0) {
                // Current config is config for tcp need for udp
                openVpnConfigDataUdp = openVpnConfigDataUdp!!
                    .replace("proto tcp", "proto udp")
                    .replace("remote $ip $tcpPort", "remote $ip $udpPort")
            }
            if (instance!!.dataUtil!!.getBooleanSetting(DataUtil.USE_DOMAIN_TO_CONNECT, false)) {
                val dDomain = "$hostName.opengw.net"
                openVpnConfigDataUdp = openVpnConfigDataUdp!!.replace(ip!!, dDomain)
            }
            // Current config is udp only
            return openVpnConfigDataUdp
        }

    val calculateSpeed: String
        get() = round(speed.toDouble() / (1000 * 1000))

    val calculateTotalTraffic: String
        get() {
            val inMB = totalTraffic.toDouble() / (1000 * 1000)
            if (inMB < 1000) {
                return round(inMB) + " MB"
            }
            val inGB = inMB / 1000
            if (inGB < 1000) {
                return round(inMB / 1000) + " GB"
            }
            return round(inGB / 1000) + " TB"
        }

    fun getCalculateUpTime(context: Context): String {
        //Display as second
        if (uptime < 60000) {
            return round((uptime / 1000).toDouble()) + " " + context.resources.getString(R.string.seconds)
        }
        //Display as minute
        if (uptime < 3600000) {
            return Math.round(uptime.toDouble() / 60000)
                .toString() + " " + context.resources.getString(R.string.minutes)
        }
        //Display as hours
        if (uptime < 3600000 * 24) {
            return round(uptime.toDouble() / 3600000) + " " + context.resources.getString(R.string.hours)
        }
        return round(uptime.toDouble() / (24 * 3600000)) + " " + context.resources.getString(R.string.days)
    }

    private fun round(value: Double): String {
        val df = DecimalFormat("####0.###")
        return df.format(value)
    }

    override fun describeContents(): Int {
        return 0
    }

    fun toVPNGateItem(): VPNGateItem {
        return VPNGateItem(
            hostName = this.hostName!!,
            ip = this.ip,
            score = this.score,
            ping = this.ping,
            speed = this.speed,
            countryLong = this.countryLong,
            countryShort = this.countryShort,
            numVpnSession = this.numVpnSession,
            uptime = this.uptime,
            totalUser = this.totalUser,
            totalTraffic = this.totalTraffic,
            logType = this.logType,
            operator = this.operator,
            message = this.message,
            openVpnConfigData = this.openVpnConfigData,
            tcpPort = this.tcpPort,
            udpPort = this.udpPort,
            isL2TPSupport = this.isL2TPSupport(),
            isSSTPSupport = this.isSSTPSupport(),
        )
    }

    fun fromVPNGateItem(vpnGateItem: VPNGateItem): VPNGateConnection {
        hostName = vpnGateItem.hostName
        ip = vpnGateItem.ip
        score = vpnGateItem.score
        ping = vpnGateItem.ping
        speed = vpnGateItem.speed
        countryLong = vpnGateItem.countryLong
        countryShort = vpnGateItem.countryShort
        numVpnSession = vpnGateItem.numVpnSession
        uptime = vpnGateItem.uptime
        totalUser = vpnGateItem.totalUser
        totalTraffic = vpnGateItem.totalTraffic
        logType = vpnGateItem.logType
        operator = vpnGateItem.operator
        message = vpnGateItem.message
        openVpnConfigData = vpnGateItem.openVpnConfigData
        tcpPort = vpnGateItem.tcpPort
        udpPort = vpnGateItem.udpPort
        isL2TPSupport = if (vpnGateItem.isL2TPSupport) 1 else 0
        isSSTPSupport = if (vpnGateItem.isSSTPSupport) 1 else 0
        return this
    }

    val name: String
        get() = this.getName(false)

    fun getName(useUdp: Boolean): String {
        var address = ip
        if (instance!!.dataUtil!!.getBooleanSetting(DataUtil.USE_DOMAIN_TO_CONNECT, false)) {
            address = "$hostName.opengw.net"
        }
        if (instance!!.dataUtil!!.getBooleanSetting(DataUtil.INCLUDE_UDP_SERVER, true)) {
            if (tcpPort == 0 && udpPort == 0) {
                // Current profile from non udp but open status page with include udp option
                return String.format("%s[%s]", countryLong, address)
            }
            return String.format(
                "%s[%s][%s]",
                countryLong,
                address,
                if (useUdp || tcpPort == 0) "UDP:$udpPort" else "TCP:$tcpPort"
            )
        }
        return String.format("%s[%s]", countryLong, address)
    }

    fun isL2TPSupport(): Boolean {
        return isL2TPSupport == 1 && Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU
    }

    fun isSSTPSupport(): Boolean {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && isSSTPSupport == 1
    }

    companion object {
        @JvmField
        val CREATOR
                : Parcelable.Creator<VPNGateConnection> =
            object : Parcelable.Creator<VPNGateConnection> {
                override fun createFromParcel(`in`: Parcel): VPNGateConnection {
                    return VPNGateConnection(`in`)
                }

                override fun newArray(size: Int): Array<VPNGateConnection?> {
                    return arrayOfNulls(size)
                }
            }

        fun fromCsv(csvLine: String): VPNGateConnection? {
            val properties =
                csvLine.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            try {
                var index = 0
                val vpnGateConnection = VPNGateConnection()
                vpnGateConnection.hostName = properties[index++]
                vpnGateConnection.ip = properties[index++]
                vpnGateConnection.score = properties[index++].toInt()
                vpnGateConnection.ping = properties[index++].toInt()
                vpnGateConnection.speed = properties[index++].toInt()
                vpnGateConnection.countryLong = properties[index++]
                vpnGateConnection.countryShort = properties[index++]
                vpnGateConnection.numVpnSession = properties[index++].toInt()
                vpnGateConnection.uptime = properties[index++].toInt()
                vpnGateConnection.totalUser = properties[index++].toInt()
                vpnGateConnection.totalTraffic = properties[index++].toLong()
                vpnGateConnection.logType = properties[index++]
                vpnGateConnection.setOperatorString(properties[index++])
                vpnGateConnection.message = properties[index++]
                vpnGateConnection.setOpenVpnConfigDataString(properties[index])
                if (instance!!.dataUtil!!.getBooleanSetting(
                        DataUtil.INCLUDE_UDP_SERVER,
                        true
                    ) && properties.size >= index + 2
                ) {
                    vpnGateConnection.tcpPort = properties[++index].toInt()
                    vpnGateConnection.udpPort = properties[++index].toInt()
                    if (properties.size > index + 1) {
                        vpnGateConnection.isL2TPSupport = properties[++index].toInt()
                    }
                    if (properties.size > index + 1) {
                        vpnGateConnection.isSSTPSupport = properties[++index].toInt()
                    }
                } else {
                    vpnGateConnection.tcpPort = 0
                    vpnGateConnection.udpPort = 0
                    vpnGateConnection.isL2TPSupport = 0
                    vpnGateConnection.isSSTPSupport = 0
                }
                return vpnGateConnection
            } catch (e: Exception) {
                return null
            }
        }
    }
}
